File: day3/code_day3/MPI/3a.bcast.f90
!======================================================================   
!  
!     example program to demonstrate Bcast
!
!======================================================================   

      program bcast

      use MPI
      
      character(12) :: message 
      integer rank
      
      integer, parameter :: root = 0 !define the root process 
      integer :: bcast_param
      
      call MPI_INIT(ierr)

      call MPI_COMM_RANK(MPI_COMM_WORLD, rank, ierr) !identify the rank
      
      if (rank .eq. root) then 
        bcast_param = 5
      else
        bcast_param = 0
      endif 
      
      call MPI_BCAST(bcast_param, 1, MPI_INT, root, &
         MPI_COMM_WORLD, ierr)
         
      write(*,*) "process", rank, ":", bcast_param
      
      call MPI_FINALIZE(ierr)
      
!======================================================================    
      end program bcast
!======================================================================    
      

File: day3/code_day3/MPI/3a.bcast.cpp
/*======================================================================   
 * 
 *     example program to demonstrate Bcast
 * 
 * ======================================================================*/   

    #include <stdio.h>
    #include <iostream>
    #include "mpi.h"

    using namespace std;

    int main(int argc, char *argv[]) {
    int rank, data;
    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    
    int root_process = 0;
    
    if (rank==0){
      data =5;
    }
    else{
      data = 10;
    }
      

    /* broadcast the value of data of rank 0 to all ranks */
    MPI_Bcast(&data, 1, MPI_INT, root_process, MPI_COMM_WORLD);

    cout << "I am rank " << rank << " and the value is " <<  data << endl;
    MPI_Finalize();
    return 0;
}